src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1125 @@
+        {
+            ChipLogDetail(
+                Zcl, "IsValidAsSupportedLocation %u - LocationName is empty string, FloorNumber, AreaType, LandmarkTag are null",
+                aLocation.locationID);
+            return false;
+        }
+    }
+
+    // If the LandmarkTag field is null, the PositionTag field SHALL be null.
+    if (aLocation.locationInfo.landmarkTag.IsNull() && !aLocation.locationInfo.positionTag.IsNull())
+    {
+        ChipLogDetail(Zcl, "IsValidAsSupportedLocation %u - PositionTag with no LandmarkTag", aLocation.locationID);
+        return false;
+    }
+
+    if (mDelegate->GetNumberOfSupportedMaps() == 0)
+    {
+        // If the SupportedMaps attribute is null, mapid SHALL be null.
+        if (!aLocation.mapID.IsNull())
+        {
+            ChipLogDetail(Zcl, "IsValidSupportedLocation %u - map Id %u is not in empty supported map list", aLocation.locationID,
+                          aLocation.mapID.Value());
+            return false;
+        }
+    }
+    else
+    {
+        // If the SupportedMaps attribute is not null, mapID SHALL be the ID of an entry from the SupportedMaps attribute.
+        if (!IsSupportedMap(aLocation.mapID.Value()))
+        {
+            ChipLogError(Zcl, "IsValidSupportedLocation %u - map Id %u is not in supported map list", aLocation.locationID,
+                         aLocation.mapID.Value());
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool Instance::IsUniqueSupportedLocation(const LocationStructureWrapper & aLocation, bool ignoreLocationId)
+{
+    BitMask<LocationStructureWrapper::IsEqualConfig> config;
+
+    if (ignoreLocationId)
+    {
+        config.Set(LocationStructureWrapper::IsEqualConfig::kIgnoreLocationId);
+    }
+
+    // If the SupportedMaps attribute is not null, each entry in this list SHALL have a unique value for the combination of the
+    // MapID and LocationInfo fields. If the SupportedMaps attribute is null, each entry in this list SHALL have a unique value for
+    // the LocationInfo field.
+    if (mDelegate->GetNumberOfSupportedMaps() == 0)
+    {
+        config.Set(LocationStructureWrapper::IsEqualConfig::kIgnoreMapId);
+    }
+
+    uint8_t locationIndex = 0;
+    LocationStructureWrapper entry;
+    while (mDelegate->GetSupportedLocationByIndex(locationIndex++, entry))
+    {
+        if (aLocation.IsEqual(entry, config))
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool Instance::ReportEstimatedEndTimeChange(const DataModel::Nullable<uint32_t> & aEstimatedEndTime)
+{
+    if (mEstimatedEndTime == aEstimatedEndTime)
+    {
+        return false;
+    }
+
+    // The value of this attribute SHALL only be reported in the following cases:
+    // - when it changes from null to any non-zero value
+    if (mEstimatedEndTime.IsNull())
+    {
+        return (aEstimatedEndTime.Value() != 0);
```

Comment:
This does not seem to match spec.  Spec says to report if changing to 0 from any value, no?

Reply:
This is due to the same misunderstanding as the previous one. Let's see what Petru says the correct interpretation is.

Reply:
Fixed [here](https://github.com/project-chip/connectedhomeip/pull/33991/commits/af7d1ba17653db04d0fa89b72ade5fb8d0101088).